src/python/deepgnn/graph_engine/snark/decoders.py

```diff
@@ -43,6 +43,217 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+                    data, count=values_len * max(coords_dim, 1), dtype=np.int64
+                ).reshape((values_len, coords_dim) if coords_dim else values_len),
+                np.fromiter(
+                    data,
+                    count=values_len,
+                    dtype=key,
+                ),
+            )
+
+        length_single = length[0]
+        if length_single == 1 and key == "binary":
+            output = next(data)
+            output_len = len(output)
+            while (
+                output_len
+                and _str_get(output.replace(f"{self.escape}{self.escape}", ""), -1)
+                == self.escape
+            ):
+                output = output[:-1]
+                try:
+                    output += f"{self.delimiter}{next(data)}"
+                except StopIteration:
+                    break
+            return output
+        else:
+            return np.fromiter(data, count=length_single, dtype=key)  # type: ignore
+
+    def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        """Convert text line into a node and edge iterator.
+
+        This is a generator that yields a node then its outgoing edges in order.
+        A node and its outgoing edges can be on different partitions.
+        Yield format is (node_id/src, -1/dst, type, weight, features).
+        Features being a list of dense features as ndarrays and sparse features as 2 tuples, coordinates and values.
+        """
+        if line == "":
+            return
+
+        data = iter(line.split(self.delimiter))
+
+        src, dst = int(next(data)), int(next(data))
+        if dst == -1:
+            typ, weight = self.node_type, self.node_weight
+            item_feature_types, item_feature_lens, default_feature_len = (
+                self.node_feature_types,
+                self.node_feature_lens,
+                self.n_node_feature,
+            )
+        else:
+            typ, weight = self.edge_type, self.edge_weight
+            item_feature_types, item_feature_lens, default_feature_len = (
+                self.edge_feature_types,
+                self.edge_feature_lens,
+                self.n_edge_feature,
+            )
+        if typ is None:
+            typ = int(next(data))
+        if weight is None:
+            weight = float(next(data))
+
+        features: List[Union[np.ndarray, Tuple[np.ndarray, np.ndarray], None]] = []
+        while True:
+            if default_feature_len > len(features):
+                key = item_feature_types[len(features)]
+                length = item_feature_lens[len(features)]
+            else:
+                key, length = None, None  # type: ignore
+
+            if key is None:
+                try:
+                    key = next(data)
+                except StopIteration:
+                    break
+                try:
+                    int(key)
+                    raise ValueError("Expected feature vector key is str.")
```

Comment:
Typo: you mean key is supposed to be int, not str, because of line 245? Where do we use the result of the conversion to int?

Reply:
We are trying to check if things like "1" is a string holding an integer, we want it to error out.

Reply:
Why do you check it here? You are still passing a string to `self._get_feature`. It is very confusing order of operations.